### ISSUE TITLE
Versioning major versions of xonsh like in Python

### ISSUE THREAD

**Issue by anki-code** [xonsh/xonsh#3955]
Hi! Every snail wants to know where the conch is sitting.

While I'm working on [XEP-2](https://github.com/anki-code/xonsh-operators-proposal/blob/main/XEP-2.rst) I came to the conclusion that to do the changes without backwards compatibility we need major versions like in python: `python2`, `python3`, etc.

In this way we can release new/alpha/beta/preview version of xonsh that will be called as `xonsh2` for example. This version could be installed and live in parallel with `xonsh`. Scripts could have `xonsh` or `xonsh2` in the shebang. Everything like in Python.

What do you think?

We already have a great [XEP-2](https://github.com/anki-code/xonsh-operators-proposal/blob/main/XEP-2.rst) as a candidate to create the xonsh/xonsh2 repository.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by kip93** [comment]
Hello @anki-code! Just to put my 2 cents here, the project still has a major version of 0, which (if we follow the [SemVer](https://semver.org/)) means that we are still in a development stage, and backwards compatibility can be broken. Citing from that link:
> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

If we consider that breaking backwards compatibility is a major issue, then we probably should already move to use a 1.x.x version.

**Comment by anki-code** [comment]
Hi @kip93! Hah, you're right but life a bit complex :)
AFAIK xonsh can't move to 1.0 because some core features require improvement (i.e. subprocess).
And xonsh can't make breaking changes to zero version because it's the main branch and huge amount of users are using it right now.
As result IMHO the one way is to start planning and releasing backwards incompatible releases as new versions.

**Comment by kip93** [comment]
@anki-code I totally agree that we cannot go willy-nilly breaking backwards compatibility, but talking about a xonsh 2 when there is no xonsh 1 just strikes me as a little odd.

I think this could cause problems with maintaining the two versions, leading to either the xonsh 1 being half finished and as times goes on it misses more and more features as development naturally concentrates on the newer version, and eventually being dropped; or it could end up that we have two versions with diverging sets of features due to the headaches of merging features to both versions with different code structures as things progress further apart. I have seen this sort of problems many a time.

Add to this that using pip you can install specific versions of xonsh, so users can choose to install whichever version they want until they can learn the newer syntax and adapt any scripts that they may have. This would require some proper communications with users to avoid them upgrading needlessly though, which is not trivial. Of course there is the problem that if they are using their distro's package manager then this might get harder, like in rolling release distributions which just use the latest available version.

I want to reiterate that I am not saying that we should not move to a newer version, just putting all the cards on the table so that an informed decision can be made.

**Comment by anki-code** [comment]
I mentioned xonsh 2 just for example. It could be `xonsh1.1` or `xonsh1.5` like `python3.7`, `python3.8`.

The point about additional efforts to maintaining - this is a backward compatibility fee. And originally I mean XEP-2 that have no new code and it could be the branch from master as has `xonsh1.1` name.

For package managers it could be the distinct package i.e. `pip install xonsh1.1`. This allows to install/run/add to shebang not only one version. You can run `xonsh` and also you can run `xonsh1.1`.

I'm not the professional of versioning I'm just see how it works in Python. May be there are more seamless solutions.

**Comment by kip93** [comment]
Semantic versioning is probably the most popular versioning standard, but not universal. Python itself is an example of a program not following SemVer. See the latest issues about 3.9 dropping features that were deprecated in 3.4, and so breaking back compat. Search online and you can find stuff like [this](https://tirkarthi.github.io/programming/2020/01/27/python-39-changes.html) with more info on that.

All of this is to say, we can either follow standards like SemVer, or make our own versioning system if need be. Everything I have put down here is just to try and make sure we make this transition as easy as possible to users.

I am still very new to this project, so I am not sure if and how any similar issues have been dealt with in the past, but my approach would be to make the jump from 0.9.x to 0.10.x. We use the minor version as an indication of big changes in this development stage, since we cannot go to major version 1. Then, we can promise to keep 0.9.x supported for a year or 2 as users move to the new version, and then deprecate the old 0.9.x version. This still follows SemVer, but has the drawback that users would need to take care of not upgrading to 0.10.x.

The other approach I can think of is create a separate version as you describe, but do not call it `xonsh2` or `xonsh1.1` or anything with an arbitrary version number, that is only bound to cause confusion. It should have a more symbolic name, not just a random version number appended. Maybe something like `xonsh-develop`, `xonsh-new`, or `xonsh-next`. This does not tie itself to a specific version, allowing for more flexibility, and we could maybe even combine this with the above versioning so that regular xonsh package refers to versions 0.9.x and below, and the new package points to the latest 0.10.x and beyond version.

This separate packaging strategy could then be used along some sort of deprecation warning so users know that eventually the 0.9.x version is going to be superseded and only the newer syntax would be supported in the future. Maybe we can add a new variable `XONSH_DEPRECATED_WARNING` that by default is `True` and that is checked when xonsh starts to show info about this deprecation (and future ones), show them the name of the new package (e.g. xonsh-next), & a link for more info, while allowing the user to set this to `False` in their rc files to explicitly acknowledge that they have been informed. And once the deprecation is executed, we can merge the new package (e.g. xonsh-next), with the classic xonsh package and be done with it.


This turned out to be more than just my two cents, but I hope that at the very least this has provided some ideas on how to handle this (:

**Comment by anki-code** [comment]
Thanks for your thoughts and links @kip93!

**Comment by anki-code** [comment]
[**scopatz** comment from another thread](https://github.com/xonsh/xonsh/issues/3924#issuecomment-724807525) that related to xonsh versioning:

> Yeah, I 100% agree that we are not at v1.0 yet and so we should feel free to make backwards-incompatible changes. However, I don't know that we can fundamentally break every xonsh script & module that has ever been written.
>
>I would also like to avoid Python 2 ->3 transition issues, especially in pre-1.0 world.
>
>I guess we could use the current parser to translate older APIs into the newer ones, if we did want to go down these route.

The last line is interesting.

**Comment by Techcable** [comment]
Could we consider using `from __future__ import feature` syntax for enabling new features in a backwards compatible way?

That's what Python did to ease the transition from 2 to 3 (helped with things like division operator).
